airbyte-db/lib/src/main/java/io/airbyte/db/factory/DataSourceFactory.java

```diff
@@ -0,0 +1,251 @@
+                                  final String password,
+                                  final String host,
+                                  final int port,
+                                  final String database,
+                                  final String driverClassName,
+                                  final Map<String, String> connectionProperties) {
+    return new DataSourceBuilder()
+        .withConnectionProperties(connectionProperties)
+        .withDatabase(database)
+        .withDriverClassName(driverClassName)
+        .withHost(host)
+        .withPort(port)
+        .withPassword(password)
+        .withUsername(username)
+        .build();
+  }
+
+  /**
+   * Convenience method that constructs a new {@link DataSource} for a PostgreSQL database using the
+   * provided configuration.
+   *
+   * @param username The username of the database user.
+   * @param password The password of the database user.
+   * @param host The host address of the database.
+   * @param port The port of the database.
+   * @param database The name of the database.
+   * @return The configured {@link DataSource}.
+   */
+  public static DataSource createPostgres(final String username,
+                                          final String password,
+                                          final String host,
+                                          final int port,
+                                          final String database) {
+    return new DataSourceBuilder()
+        .withDatabase(database)
+        .withDriverClassName("org.postgresql.Driver")
+        .withHost(host)
+        .withPort(port)
+        .withPassword(password)
+        .withUsername(username)
+        .build();
+  }
+
+  /**
+   * Builder class used to configure and construct {@link DataSource} instances.
+   */
+  private static class DataSourceBuilder {
+
+    private static final Map<String, String> JDBC_URL_FORMATS = Map.of("org.postgresql.Driver", "jdbc:postgresql://%s:%d/%s",
+        "com.amazon.redshift.jdbc.Driver", "jdbc:redshift://%s:%d/%s",
+        "com.mysql.cj.jdbc.Driver", "jdbc:mysql://%s:%d/%s",
+        "com.microsoft.sqlserver.jdbc.SQLServerDriver", "jdbc:sqlserver://%s:%d/%s",
+        "oracle.jdbc.OracleDriver", "jdbc:oracle:thin:@%s:%d:%s",
+        "ru.yandex.clickhouse.ClickHouseDriver", "jdbc:ch://%s:%d/%s",
+        "org.mariadb.jdbc.Driver", "jdbc:mariadb://%s:%d/%s");
+
+    private Map<String, String> connectionProperties = Map.of();
+    private String database;
+    private String driverClassName = "org.postgresql.Driver";
+    private String host;
+    private String jdbcUrl;
+    private Integer maximumPoolSize = 5;
+    private Integer minimumPoolSize = 0;
+    private String password;
+    private Integer port = 5432;
+    private String username;
+
+    private DataSourceBuilder() {}
+
+    public DataSourceBuilder withConnectionProperties(final Map<String, String> connectionProperties) {
```

Comment:
it is already written, but the `@Builder` annotation of lombok can help you writing builders without having to implements them yourself.